core/embed/rust/src/ui/layout_delizia/component/keyboard/word_count.rs

```diff
@@ -4,91 +4,110 @@ use crate::ui::{
+pub struct SelectWordCount {
+    keypad: ValueKeypad,
 }
 
-impl SelectWordCount {
-    pub fn new_all() -> Self {
-        Self::All(ValueKeypad::new())
-    }
+type Label = &'static str;
 
-    pub fn new_multishare() -> Self {
-        Self::Multishare(BinarySelection::new(
-            ButtonContent::Text("20".into()),
-            ButtonContent::Text("33".into()),
-            theme::button_keyboard(),
-            theme::button_keyboard(),
-        ))
+impl SelectWordCount {
+    pub fn new(choices: Vec<u32, 5>, labels: Vec<Label, 5>) -> Self {
+        Self {
+            keypad: ValueKeypad::new(choices, labels),
+        }
     }
 }
 
 impl Component for SelectWordCount {
     type Msg = SelectWordCountMsg;
 
     fn place(&mut self, bounds: Rect) -> Rect {
-        match self {
-            SelectWordCount::All(full_selector) => full_selector.place(bounds),
-            SelectWordCount::Multishare(bin_selector) => bin_selector.place(bounds),
-        }
+        self.keypad.place(bounds)
     }
 
     fn event(&mut self, ctx: &mut EventCtx, event: Event) -> Option<Self::Msg> {
-        match self {
-            SelectWordCount::All(full_selector) => full_selector.event(ctx, event),
-            SelectWordCount::Multishare(bin_selector) => {
-                if let Some(m) = bin_selector.event(ctx, event) {
-                    return match m {
-                        BinarySelectionMsg::Left => Some(SelectWordCountMsg::Selected(20)),
-                        BinarySelectionMsg::Right => Some(SelectWordCountMsg::Selected(33)),
-                    };
-                }
-                None
-            }
-        }
+        self.keypad.event(ctx, event)
     }
 
     fn render<'s>(&'s self, target: &mut impl Renderer<'s>) {
-        match self {
-            SelectWordCount::All(full_selector) => full_selector.render(target),
-            SelectWordCount::Multishare(bin_selector) => bin_selector.render(target),
-        }
+        self.keypad.render(target)
     }
 }
 
-pub struct ValueKeypad {
-    button: [Button; Self::NUMBERS.len()],
+type Cell = (usize, usize, usize, usize);
+
+struct ValueKeypad {
+    buttons: Vec<(Button, Option<u32>, Cell), 6>,
     keypad_area: Rect,
 }
 
 impl ValueKeypad {
-    const NUMBERS: [u32; 5] = [12, 18, 20, 24, 33];
-    const LABELS: [&'static str; 5] = ["12", "18", "20", "24", "33"];
-    const CELLS: [(usize, usize); 5] = [(0, 0), (0, 2), (1, 0), (1, 2), (2, 1)];
-
-    fn new() -> Self {
-        ValueKeypad {
-            button: Self::LABELS.map(|t| {
-                Button::with_text(t.into())
-                    .styled(theme::button_keyboard())
-                    .with_text_align(Alignment::Center)
-            }),
+    /*
+     * 0 | 1
+     * -----
+     * 2 | 3
+     * -----
+     * x | 4
+     */
+    const SIX_CELLS: [Cell; 6] = [
+        (0, 0, 1, 1),
+        (0, 2, 1, 1),
+        (2, 0, 1, 1),
+        (2, 2, 1, 1),
+        (4, 0, 1, 1),
+        (4, 2, 1, 1),
```

Comment:
_here_, you can leave an empty space in `SIX_CELLS` for the cell in bottom left, and make a special field for `cancel_cell` that you fill out -- either with the big one, or with the bottom left one. this again allows you to zip the buttons with `CELLS` in `place()`, except now `cancel` will be its own button, not part of the vec.

that also means that `buttons` don't need an `Option` for the value